docs/core.md

```diff
@@ -0,0 +1,361 @@
+# Core library
+
+The *Core library* implements common assertions about the exit code and output
+captured with the most recent `run()` command.
+
+
+## `flunk()`
+
+Display an error message and fail. This function provides a convenient way to
+report failure in arbitrary situations. You can use it to implement your own
+helpers when the ones available do not meet your needs. Other functions use it
+internally as well.
+
+```bash
+@test 'flunk()' {
+  flunk 'this test always fails'
+}
+```
+
+The message can also be specified on the standard input.
+
+```bash
+@test 'flunk() with pipe' {
+  echo 'this test always fails' | flunk
```

Comment:
Piping simplifies the design and implementation of how output is generated by library functions.

When a test fails, output is generated in stages. Each stage has a well-defined job and pipes its output into the next stage.

1. **Generate details**: `{ ... }` groups together the output of multiple commands that produce the formatted details relevant to the failed test, which is then all piped into the next stage at once.
2. **Add header and footer**: `_decorate` adds the header and footer around the input received from the previous stage.
3. **Display and fail**: `flunk` redirects input to `STDERR` and returns 1, effectively causing the test to fail and the output to be displayed.

This is cleaner than the alternative of separately printing a header, body and footer, all while having all function individually redirect output to `STDERR`.

See [`assert_success`](https://github.com/sstephenson/bats/pull/110/files#diff-64fb509098cd5357d135c130ca2176a2R122) or [`assert_line`](https://github.com/sstephenson/bats/pull/110/files#diff-64fb509098cd5357d135c130ca2176a2R199) for example.